<?xml version="1.0"?>
<robot name="futurakart" xmlns:xacro="http://ros.org/wiki/xacro">

    <!-- Include macros -->
    <xacro:include filename="$(find futurakart_description)/urdf/macros/model.urdf" />
    <xacro:include filename="$(find futurakart_description)/urdf/macros/kinect.urdf" />

    <!-- Bring in simulation data for Gazebo. -->
    <xacro:include filename="$(find futurakart_description)/urdf/futurakart.gazebo" />

    <!--**************************************************************************-->
    <!--                     Define links and joints                              -->
    <!--**************************************************************************-->

    <!-- Center of the Kart -->
    <link name="base_link"/>

    <link name="chassis_link">
        <visual>
            <geometry>
                <box size="${base_length} ${base_width} ${base_height}"/>
            </geometry>
            <material name="yellow" />
        </visual>
        <inertial>
            <!-- Center of mass -->
            <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
            <mass value="16.523"/>
            <!-- Moments of inertia: ( chassis without wheels ) -->
            <inertia
                    ixx="0.3136" ixy="-0.0008" ixz="0.0164"
                    iyy="0.3922" iyz="-0.0009"
                    izz="0.4485"/>
        </inertial>
    </link>
    <joint name="chassis_link_joint" type="fixed">
        <parent link="base_link"/>
        <child link="chassis_link"/>
        <origin xyz="-${wheelbase/2} 0 ${wheel_radius}"/>
    </joint>


    <!-- Center of the Kart projected on the floor-->
    <link name="base_footprint" />
    <joint name="base_footprint_link_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_footprint"/>
        <origin xyz="-${wheelbase/2} 0 -${vertical_offset}"/>
    </joint>


    <!--**************************************************************************-->
    <!--Define 4 wheels-->
    <!--**************************************************************************-->

    <xacro:front_wheel prefix="left">
        <!--Steering arm position-->
        <origin xyz="0 ${-rot_axle_dist/2} ${wheel_vertical_offset}" rpy="0 0 0" />
        <origin xyz="0 ${-track_arm} 0" rpy="0 0 0" />
    </xacro:front_wheel>

    <xacro:front_wheel prefix="right">
        <!--Steering arm position-->
        <origin xyz="0 ${rot_axle_dist/2} ${wheel_vertical_offset}" rpy="0 0 0" />
        <origin xyz="0 ${track_arm} 0" rpy="0 0 0" />
    </xacro:front_wheel>

    <!--<xacro:wheel prefix="front_right">-->
        <!--<origin xyz="0 ${rot_axle_dist/2} ${wheel_vertical_offset}" rpy="0 0 0" />-->
        <!--<axis xyz="0 0 1" />-->
        <!--&lt;!&ndash;For Gazebo only&ndash;&gt;-->
        <!--<transmission name="front_right_wheel_trans">-->
            <!--<type>transmission_interface/SimpleTransmission</type>-->
            <!--<joint name="front_right_wheel">-->
                <!--&lt;!&ndash;<hardwareInterface>PositionJointInterface</hardwareInterface>&ndash;&gt;-->
                <!--<hardwareInterface>VelocityJointInterface</hardwareInterface>-->
            <!--</joint>-->
            <!--<actuator name="front_right_actuator">-->
                <!--<mechanicalReduction>1</mechanicalReduction>-->
            <!--</actuator>-->
        <!--</transmission>-->
    <!--</xacro:wheel>-->

    <xacro:rear_wheel prefix="left">
        <origin xyz="-${wheelbase} ${-track_rear/2} ${wheel_vertical_offset}" rpy="0 0 0" />
    </xacro:rear_wheel>

    <xacro:rear_wheel prefix="right">
        <origin xyz="-${wheelbase} ${track_rear/2} ${wheel_vertical_offset}" rpy="0 0 0" />
    </xacro:rear_wheel>


    <!--**************************************************************************-->
    <!--Define sensors -->
    <!--**************************************************************************-->
    <xacro:sensor_kinect parent="base_link"/>

</robot>
